_episodes/41-code-review.md

```diff
@@ -465,6 +476,72 @@ and [Smartbear](https://smartbear.com/learn/code-review/best-practices-for-peer-
+reformatted, refactored and changed the behavior of the code make sure each of these
+is in a separate commit (i.e reformat the code, commit, refactor the code, commit, alter
+the behavior of the code, commit).
+
+Make sure you write a clear description of the content and purpose of the change.
+This should be provided as the pull request description.
+This should provide the context that reading the code will make more sense.
+
+It is also a good idea to review your code yourself. In doing this you will spot
+the more obvious issues with your code, allowing your reviewer to focus on the
+things you cannot spot.
+
+## Empathy in review comments
+
+Code is written by humans (mostly!), and code review is a form of communication. As such
+empathy is really important for effective reviewing.
+
+When reviewing code, it can be sometimes frustrating when code is confusing, particularly
+as it will be implemented differently to how you would have done it. However, it is important
+as a reviewer to be compassionate to the person whose code you are reviewing. Specifically:
+
+* Identify positives in code as and when you find them (particularly if it is an improvement on
+something you've previously fed back on in a previous review).
+* Remember different doesn't mean better - only request changes if the code is wrong or hard to understand.
+* Limit the number of non-critical suggestions to a few - you are aiming for better rather than perfect.
+* Ask questions to understand why something has been done a certain way rather than assuming you
+know a better way.
+* If a conversation is taking place on a review and hasn't been resolved by a
+single back-and-forth exchange, then schedule a conversation to discuss instead
+(recording the result on the PR).
+
+## Designing a review process
+
+To maximize the benefit of a code review it needs a process that is followed by everyone
+involved, and that everyone believes that process provides value.
+
+One way to foster this is to design the process as a team. When you're doing this you
+should consider:
+
+* Do all changes need to go through code review
+* What technologies will you use to manage the review process
+* How quickly do you expect someone to review the code once you've raised a PR?
+* How long should be spent reviewing code?
+* What kind of issues are (and aren't) appropriate to raise in a PR?
+* How will someone know when they are expected to take action (e.g. review a PR).
+
+Once you've introduced a review process, you should monitor (either formally or
+informally) how well it is working.
+
+It is important that reviews are processed quickly, to avoid costly context switching.
+We recommend aiming for 3 hours to get a first review, with the PR being merged the same
+day in most cases. If you are regularly missing these targets, then you should review
+where things are getting stuck and work out what you can do to move things along.
```

Comment:
Here I am assuming the reader is part of a team. However, I remember in discussions we talked how lots of users of this course are working independently. How do those scientists who are working independently use the code review section? Do they find someone else in their dept to review their code, reviewing it themselves, or just skipping this section?

Reply:
This is a good question. As someone who has lone coded in various scientific settings, I still use
merge/pull requests as a gate check on what I have just done. Of course, I haven't done that in
every instance, but it is a nice idea to strive for.

In terms of actual code review, this is tough. I was part of a community effort to try and define
what code review in research might look like: https://dev-review.readthedocs.io/en/latest/index.html

There, we did recommend trying to find someone to review your code. That person could be in your
research group (but just not working directly on what you are). We could link to that page here do
deal with the case of solo researchers working on code.